core/src/main/java/org/infinispan/topology/ClusterTopologyManagerImpl.java

```diff
@@ -460,15 +460,17 @@ private void waitForJoinerView(Address joiner, int joinerViewId, long timeout)
                log.tracef("Waiting to recover cluster status before processing join request from %s", joiner);
             }
          }
-         long endTime = timeService.expectedEndTime(timeout, TimeUnit.MILLISECONDS);
+         final long endTime = timeService.expectedEndTime(timeout, TimeUnit.MILLISECONDS);
+         long remainingTime = -1;
          synchronized (clusterManagerLock) {
             while (viewId < joinerViewId || clusterManagerStatus == ClusterManagerStatus.RECOVERING_CLUSTER) {
-               if (timeService.isTimeExpired(endTime) || !clusterManagerStatus.isRunning())
+               remainingTime = timeService.remainingTime(endTime, TimeUnit.MILLISECONDS);
+               if (remainingTime == 0 || !clusterManagerStatus.isRunning())
                   break;
-               clusterManagerLock.wait(timeService.remainingTime(endTime, TimeUnit.MILLISECONDS));
+               clusterManagerLock.wait(remainingTime);
             }
          }
-         if (timeService.isTimeExpired(endTime)) {
+         if (remainingTime == 0) { //zero is only returned by TimeService.remainingTime() when it expires.
```

Comment:
Would it be possible to move this throw into the while loop (and rearrange lines 468-470)? This way it would be a little bit more readable.